packages/core/src/utils/oidc-provider-event-listener.ts

```diff
@@ -39,13 +42,35 @@ export const grantSuccessListener = async (
     id_token && 'idToken',
   ].filter((value): value is IssuedTokenType => notFalsy(value));
 
-  const grantType = params?.grant_type;
-  const type: LogType =
-    grantType === GrantType.AuthorizationCode ? 'CodeExchangeToken' : 'RefreshTokenExchangeToken';
+  const grantType: unknown = params?.grant_type;
+  const type = getLogType(grantType as GrantType);
   ctx.log(type, {
     userId: account?.accountId,
     params,
     issued,
     scope,
   });
 };
+
+export const grantErrorListener = async (
+  ctx: KoaContextWithOIDC & WithLogContext & { body: GrantBody },
+  error: errors.OIDCProviderError
+) => {
+  const {
+    oidc: {
+      entities: { Client: client },
+      params,
+    },
+  } = ctx;
+  ctx.addLogContext({
+    applicationId: client?.clientId,
+  });
+
+  const grantType: unknown = params?.grant_type;
+  const type = getLogType(grantType as GrantType);
```

Comment:
Nope, as grantType can be `undefined`, so it at least should be `Optional<GrantType>`

Can you try something like this:
```typescript
const type = getLogType(params?.grant_type as Optional<GrantType>);
```

And in the static method, define grantType as an optional param
```typescript
const getLogType = (grantType?: GrantType): LogType => ...

Reply:
Refactored. PTAL

Reply:
nit: Add a colon to the end of the error message

Reply:
Updated and add new UTs of skipping logging. PTAL

```ts
    console.error('Unexpected grant_type:', grantType);
```